docs/practical-examples/forms.md

```diff
@@ -0,0 +1,168 @@
+
+💡 To understand how `FormBuilder` works, it's a good idea to see how the [default class](https://github.com/rails/rails/blob/main/actionview/lib/action_view/helpers/form_helper.rb){:target="_blank"} in Rails defines helper
+
+```ruby
+class YourCustomFormBuilder < ActionView::Helpers::FormBuilder
+  def text_area_with_label(attribute, **options)
+    @template.capture do
+      @template.concat label(attribute, options.merge(class: 'your-css-classes')
+      @template.concat text_area(attribute, options.merge(class: 'your-css-classes')
+    end
+  end
+end
+```
+👀 Note that you need to define the `builder` that `form_with` should use
+
+```erb
+<!-- app/components/comment_form_component.html.erb -->
+<%= form_with(model: comment, builder: YourCustomFormBuilder) do |form| %>
+  <%= form.text_area_with_label :content %>
+<% end %>
+```
+
+### ✅ Best: Abstract the markup into a `ViewComponent`
+
+In order to clean up your `FormBuilder` and be able to test the markup in isolation, you can extract the code inside the form builder method into a `ViewComponent`. This is very similar to what [Rails' internals](https://github.com/rails/rails/blob/d3f4db9e95a475822c02b29241a5d07cbcff6fd5/actionview/lib/action_view/helpers/form_helper.rb#L1151){:target="_blank"} do to create it's default form helpers.
+
+Let's extract the code from the previous example into a `ViewComponent`
+
+```ruby
+class YourCustomFormBuilder < ActionView::Helpers::FormBuilder
+  # Note: (attribute, **options) is an oversimplification of the arguments passed into
+  # a form builder method and a ViewComponent. Take a look at the links above to understand
+  # how default form builder methods work, which arguments they receive, etc.
+  def text_area_with_label(attribute, **options)
+    @template.render TextAreaWithLabelComponent.new(attribute: attribute, **options)
+  end
+end
+```
+
+```ruby
+class TextAreaWithLabelComponent < ApplicationComponent
+  def initialize(attribute:, **options)
+    @attribute = attribute
+  end
+  attr_reader :attribute
+end
+```
+
+```erb
+<!-- app/components/text_area_with_label_component.html.erb -->
+<div class='maybe-some-wrapper-class'>
+  label_tag attribute, class: 'your-label-css-classes'
+  text_area_tag attribute, class: 'your-text-area-classes'
+</div>
+```
+
+```erb
+<!-- app/components/comment_form_component.html.erb -->
+<%= form_with(model: comment, builder: YourCustomFormBuilder) do |form| %>
+  <%= form.text_area_with_label :content %>
+<% end %>
+```
+
+#### Writing input elements in vanilla HTML
+
+Nothing is really stopping you from using regular `<input>` tags instead of Rails' helper methods. However, is important for your own implementations to follow the conventions set by other form helpers. If you plan to do so, please read the [Understanding Parameter Naming Conventions](https://guides.rubyonrails.org/form_helpers.html#understanding-parameter-naming-conventions){:target="_blank"} section of the `form_helpers` Rails guide.
```

Comment:
```suggestion
Nothing is really stopping you from using regular `<input>` tags instead of Rails' helper methods. However, it is important for your own implementations to follow the conventions set by other form helpers. If you plan to do so, please read the [Understanding Parameter Naming Conventions](https://guides.rubyonrails.org/form_helpers.html#understanding-parameter-naming-conventions){:target="_blank"} section of the `form_helpers` Rails guide.
```